src/common/base/ICord.h

```diff
@@ -0,0 +1,247 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#pragma once
+
+#include <base/Base.h>
+
+namespace nebula {
+
+template <std::size_t kBlockContentSize = 1024>
+class ICord {
+public:
+    static_assert(kBlockContentSize && !(kBlockContentSize & (kBlockContentSize-1)),
+                  "kBlockContentSize must be power of 2");
+    ICord() : head_(inlineBlock_), tail_(head_) {}
+
+    virtual ~ICord() {
+        clear();
+    }
+
+    size_t size() const noexcept {
+        return len_;
+    }
+
+    bool empty() const noexcept {
+        return len_ == 0;
+    }
+
+    void clear() {
+        auto alloc = next(head_);
+        if (alloc) {
+            DCHECK(tail_);
+
+            // Need to release all blocks
+            char* p = alloc;
+            while (p != tail_) {
+                char* n = next(p);
+                free(p);
+                p = n;
+            }
+            // Free the last block
+            free(p);
+        }
+
+        len_ = 0;
+
+        head_ = nullptr;
+        tail_ = nullptr;
+    }
+
+    // Apply each block to the visitor until the end or the visitor
+    // returns false
+    bool applyTo(std::function<bool(const char*, int32_t)> visitor) const {
+        if (empty()) {
+            return true;
+        }
+
+        char* n = head_;
+        while (n != tail_) {
+            if (!visitor(n, kBlockContentSize)) {
+                // stop visiting further
+                return false;
+            }
+            // Get the pointer to the next block
+            n = next(n);
+        }
+
+        // Last block
+        return visitor(tail_, lengthMod());
+    }
+
+    // Append the cord content to the given string
+    size_t appendTo(std::string& str) const {
+        if (empty()) {
+            return 0;
+        }
+
+        char* n = head_;
+        while (n != tail_) {
+            str.append(n, kBlockContentSize);
+            // Get the pointer to the next block
+            n = next(n);
+        }
+
+        // Last block
+        str.append(tail_,
+                   lengthMod() == 0 ? kBlockContentSize : lengthMod());
+
+        return len_;
+    }
+
+    // Convert the cord content to a new string
+    std::string str() const {
+        std::string buf;
+        buf.reserve(len_);
+        appendTo(buf);
+
+        return buf;
+    }
+
+    ICord<kBlockContentSize>& write(const char* value, size_t len) {
+        if (len == 0) {
+            return *this;
+        }
+
+        size_t bytesToWrite =
```

Comment:
It seems you calculate the lengthMod multi times, such as.
1.  line 110
2.  inside isFull
3. line 115


Could we unify them?

Reply:
Fixed.